Assets/MixedRealityToolkit/Services/BaseEventSystem.cs

```diff
@@ -25,44 +39,226 @@ public abstract class BaseEventSystem : BaseService, IMixedRealityEventSystem
         public virtual void HandleEvent<T>(BaseEventData eventData, ExecuteEvents.EventFunction<T> eventHandler) where T : IEventSystemHandler
         {
             Debug.Assert(!eventData.used);
+
             eventExecutionDepth++;
 
+            // First handle events via old API part, via GameObjects.
+            // This sends event to every component of matching type on this object, 
+            // INCLUDING the ones that don't expect it.
             for (int i = EventListeners.Count - 1; i >= 0; i--)
             {
                 ExecuteEvents.Execute(EventListeners[i], eventData, eventHandler);
             }
 
+            // Send events to all handlers registered via RegisterHandler API.
+            HashSet<IEventSystemHandler> handlers;
+            if (EventHandlersByType.TryGetValue(typeof(T), out handlers))
+            {
+                foreach (IEventSystemHandler handler in handlers)
```

Comment:
The only thing I'd be careful of here is GC created by using foreach over this enumeration. I know that historically foreach has been slower than for (probably fixed in some .NET or something else) and that this would also actually trigger alloc (because it would allocate an enumerator in order to do the foreach).

It would be worth doing a quick profile of this using the Unity profiler to verify that this doesn't do per-frame GC alloc (i.e. a one time thing is fine, something on every frame or something that happens continuously on every N frames is bad). This is especially interesting because these events CAN be noisy (i.e. joint updates happen each frame) so we basically need to consider this to be happening at a rate Nx of the update loop.

Reply:
If this IS happening, we might want to snap a list<> upon each register/unregister call, so that we can for() instead of foreach()

Reply:
I don't think I can iterate over HashSet without creating an enumerator. I can do this manually by `GetEnumerator` + `MoveNext`, but it might have the same GC issue.

Reply:
Somehow I wasn't shown the second part of the comment. Yeah, list should work here.